add ci test coverage

Radwa85 opened the pull request:
1- Configured Kover to generate test coverage reports.

2- Created a GitHub Actions workflow that automatically runs tests and generates a coverage report on every pull request to the development branch.

3-Added logic to block PRs if overall coverage is below 80%

Radwa85 committed:
add gitHub action for test coverage check

Radwa85 committed:
add JaCoCo for code coverage analysis

Radwa85 committed:
rename build job to test-coverage in ci workflow

Radwa85 committed:
disable XML report for test coverage

Radwa85 committed:
fix Jacoco report path for github actions

Radwa85 committed:
enable XML report for JaCoCo test coverage

Radwa85 committed:
verify test tasks in ci workflow

Radwa85 committed:
add JacocoCoverageVerification for code coverage

Radwa85 committed:
update GitHub workflow Java version to 22

Radwa85 committed:
Update Java version to 22 and Jacoco version

Radwa85 committed:
update excludes in app/build.gradle.kts

mosayed01 commented on app/build.gradle.kts:
```diff
@@ -28,15 +29,23 @@ android {
     }
     kotlinOptions {
-        jvmTarget = "11"
+        jvmTarget = "22"
```
Why?

Radwa85 replied:
i thought it wasn`t compatible with jacoco but i will return it

mosayed01 commented:
The branch name violates the squad rules

Radwa85 committed:
return Java version to 11

Radwa85 committed:
rename ci workflow file

Radwa85 committed:
remove JaCoCo from debug build type

Ahmed7osny1 commented on .github/workflows/ci_test_coverage.yml:
```diff
@@ -0,0 +1,43 @@
+        run: ./gradlew :app:testDebugUnitTest :app:jacocoTestReport
+
+
+      - name: Check if coverage is >= 80%
```
you can use more nameful like Evaluate code coverage, you can change percentage in future you change here but we don't need to do it

Ahmed7osny1 commented on .github/workflows/ci_test_coverage.yml:
```diff
@@ -0,0 +1,43 @@
+          MISSED=$(grep -oPm1 "(?<=<counter type=\"LINE\" missed=\")[^\"]+"  app/build/reports/jacoco/jacocoTestReport/jacocoTestReport.xml)
+          COVERED=$(grep -oPm1 "(?<=<counter type=\"LINE\" covered=\")[^\"]+"  app/build/reports/jacoco/jacocoTestReport/jacocoTestReport.xml)
+          TOTAL=$(awk -v m=$MISSED -v c=$COVERED 'BEGIN {print m + c}')
+          PERCENT=$(awk -v m=$MISSED -v c=$COVERED 'BEGIN {print (c / (c + m)) * 100}')
```
i think you need to check about extract coverage data from report like this 
if [[ -z "$MISSED" || -z "$COVERED" ]]; then
            echo "Failed to extract coverage data from report"
            exit 1
          fi

Ahmed7osny1 commented on .github/workflows/ci_test_coverage.yml:
```diff
@@ -0,0 +1,43 @@
+          TOTAL=$(awk -v m=$MISSED -v c=$COVERED 'BEGIN {print m + c}')
+          PERCENT=$(awk -v m=$MISSED -v c=$COVERED 'BEGIN {print (c / (c + m)) * 100}')
+          echo "Line coverage is $PERCENT%"
+          THRESHOLD=80
+          RESULT=$(awk -v p=$PERCENT -v t=$THRESHOLD 'BEGIN {print (p < t)}')
+          if [[ $RESULT == 1 ]]; then
+            echo "Coverage below 80%!"
+            exit 1
+          else
+            echo "Coverage passed with $PERCENT%"
+          fi
```
you can simplify these lines by ussing
THRESHOLD=80
          if (( $(echo "$PERCENT < $THRESHOLD" | bc -l) )); then
            echo "Coverage below ${THRESHOLD}%!"
            exit 1
          else
            echo "Coverage passed: ${PERCENT}%"
          fi

Ahmed7osny1 commented on app/build.gradle.kts:
```diff
@@ -56,4 +65,79 @@ dependencies {
+tasks.register<JacocoReport>("jacocoTestReport") {
+    dependsOn("testDebugUnitTest")
+
+    val fileFilter = listOf(
+        "**/R.class", "**/R$*.class",
+        "**/BuildConfig.*",
+        "**/Manifest*.*",
+        "**/*Test*.*",
+        "android/**/*.*",
+        "sun/security/smartcardio/**",
+        "**/MainActivity.*",
+        "**/MainActivity\$*.*",
+        "**/*Activity.*",
+        "**/*\$WhenMappings.*",
+        "**/ui/theme/**",
+        "**/ComposableSingletons*.*"
+    )
```
why are these lines repeated? You can use it as Shared

Ahmed7osny1 commented on app/build.gradle.kts:
```diff
@@ -56,4 +65,79 @@ dependencies {
+tasks.register<JacocoCoverageVerification>("jacocoTestCoverageVerification") {
+    dependsOn("jacocoTestReport")
+
+    val fileFilter = listOf(
+        "**/R.class", "**/R$*.class",
+        "**/BuildConfig.*",
+        "**/Manifest*.*",
+        "**/*Test*.*",
+        "android/**/*.*",
+        "sun/security/smartcardio/**",
+        "**/MainActivity.*",
+        "**/MainActivity\$*.*",
+        "**/*Activity.*",
+        "**/*\$WhenMappings.*",
+        "**/ui/theme/**",
+        "**/ComposableSingletons*.*"
+    )
```
and here

mosayed01 commented on .github/workflows/ci_test_coverage.yml:
```diff
@@ -0,0 +1,43 @@
+
+
+      - name: Check if coverage is >= 80%
+        run: |
```
Search about how to make jacoco works for multimodule project and apply your result here.

mosayed01 commented:
Also u can use codecov github action to display the code coverage report result @Radwa85

Radwa85 committed:
simplify and remove duplication in Jacoco coverage check step

Radwa85 committed:
refactor Jacoco file filter for reuse

Radwa85 committed:
integrate Codecov for test coverage reporting

Radwa85 committed:
update Java version for test coverage workflow

Radwa85 committed:
remove Greeting composable and its preview

Radwa85 committed:
replace JaCoCo with Kover for code coverage

Radwa85 committed:
add kover to libs.versions.toml

Radwa85 committed:
replaced Jacoco with Kover and automated coverage check using koverVerify

mosayed01 commented:
@Radwa85 have u tested it?

Ahmed7osny1 commented on .github/workflows/ci_test_coverage.yml:
```diff
@@ -0,0 +1,37 @@
+      - name: Set up JDK
+        uses: actions/setup-java@v3
+        with:
+          java-version: '22'
```
Use version 17 like Aiman for more consistency

Ahmed7osny1 commented on .github/workflows/ci_test_coverage.yml:
```diff
@@ -0,0 +1,37 @@
+on:
+  pull_request:
+    branches:
+      - development
```
i think the best to check for all branches not in development only, you can use '*' for all branches

AimanYosofiUlt replied:
I think you should check main branch too, something like:
` branches: [main, development]`

Ahmed7osny1 commented:
In general, you are good, but you need to change your commit to use Kover, not Jacoco
I don't know if you have tested this in a multi-module or not.

AimanYosofiUlt commented on build.gradle.kts:
```diff
@@ -3,4 +3,35 @@ plugins {
+        reports {
+            filters {
+                excludes {
+                    classes(
```
Meybe it's better if you store the values inside a list for better readability 

```
val excludedPackages = listOf(
    "*.R",
    "*.R_*",
    "*.BuildConfig*",
    "*.Manifest*",
    "com.sanaa.novix.ui.theme.*",
    "*.ComposableSingletons*",
    "*.MainActivity*"
)

kover {
    reports {
        filters { excludes { classes(excludedPackages) } }
...
```

AimanYosofiUlt commented on build.gradle.kts:
```diff
@@ -3,4 +3,35 @@ plugins {
     alias(libs.plugins.android.application) apply false
     alias(libs.plugins.kotlin.android) apply false
     alias(libs.plugins.kotlin.compose) apply false
+    alias(libs.plugins.kover)
```
I see here you use `kover`, is it better than JCoCo?

Radwa85 commented:
> @Radwa85 have u tested it?
@mosayed01 "Yes, I switched to the modularization branch and added some classes for test I confirmed that individual coverage reports are generated for each module which indicates tests are working per module

mostafa-n3ma commented on .github/workflows/ci_test_coverage.yml:
```diff
@@ -0,0 +1,37 @@
+
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
```
use the latest checkout action version @v4

mostafa-n3ma commented on .github/workflows/ci_test_coverage.yml:
```diff
@@ -0,0 +1,37 @@
+        uses: actions/checkout@v3
+
+      - name: Set up JDK
+        uses: actions/setup-java@v3
```
@v4

mostafa-n3ma reviewed (commented):
nice work 
consider reading my coments please

Radwa85 committed:
use all project instead of sub project

Radwa85 committed:
use v4 instead of v3 , cache gradle,add main branch and use java 17

Radwa85 committed:
use v4 instead of v3 , cache gradle,add main branch and use java 17

Radwa85 committed:
use v4 instead of v3 , cache gradle,add main branch and use java 17

Radwa85 committed:
use v4 instead of v3 , cache gradle,add main branch and use java 17

Radwa85 committed:
use v4 instead of v3 , cache gradle,add main branch and use java 17

Radwa85 committed:
refactor kover excluded packages in build.gradle.kts

Radwa85 committed:
Merge remote-tracking branch 'origin/development' into task/ci-test-coverage

# Conflicts:
#	app/build.gradle.kts
#	build.gradle.kts
#	gradle/libs.versions.toml

AimanYosofiUlt commented on build.gradle.kts:
```diff
@@ -4,4 +4,88 @@ plugins {
+        }
+    }
+}
+/*
```
if this is not necessary, please remove this commented code

Radwa85 committed:
Merge remote-tracking branch 'origin/development' into task/ci-test-coverage

# Conflicts:
#	build.gradle.kts
#	gradle/libs.versions.toml

Radwa85 committed:
enable typesafe project accessors
